packages/datadog-plugin-couchbase/src/index.js

```diff
@@ -1,171 +1,91 @@
+      'component': 'couchbase',
+      'service.name': this.config.service || `${this.tracer._service}-couchbase`,
+      'resource.name': `couchbase.${operation}`,
+      'span.kind': 'client'
     }
-  }
-}
-
-function createWrapStore (tracer, config, operation) {
-  return function wrapStore (store) {
-    return function storeWithTrace (key, value, options, callback) {
-      const callbackIndex = findCallbackIndex(arguments)
-
-      if (callbackIndex < 0) return store.apply(this, arguments)
-
-      const scope = tracer.scope()
-      const span = startSpan(tracer, config, operation)
-
-      addBucketTag(span, this)
 
-      arguments[callbackIndex] = wrapCallback(span, arguments[callbackIndex])
-
-      return scope.bind(store, span).apply(this, arguments)
+    for (const tag in customTags) {
+      tags[tag] = customTags[tag]
     }
-  }
-}
+    const span = this.tracer.startSpan(`couchbase.${operation}`, {
+      childOf: store ? store.span : null,
+      tags
+    })
 
-function addBucketTag (span, bucket) {
-  span.setTag('couchbase.bucket.name', bucket.name || bucket._name)
-}
+    span.setTag('couchbase.bucket.name', bucket.name || bucket._name)
 
-function findCallbackIndex (args) {
-  for (let i = args.length - 1; i >= 2; i--) {
-    if (typeof args[i] === 'function') return i
+    analyticsSampler.sample(span, this.config.measured)
+    return span
   }
 
-  return -1
+  constructor (...args) {
+    super(...args)
+
+    this.addSubs('_n1qlReq', ([resource, bucket]) => {
+      const store = storage.getStore()
+      const span = this.startSpan('query', { 'span.type': 'sql', 'resource.name': resource }, store, bucket)
+      this.enter(span, store)
+    })
+
+    this.addSubs('upsert', ([bucket]) => {
+      const store = storage.getStore()
+      const span = this.startSpan('upsert', {}, store, bucket)
+      this.enter(span, store)
+    })
+
+    this.addSubs('insert', ([bucket]) => {
+      const store = storage.getStore()
+      const span = this.startSpan('insert', {}, store, bucket)
+      this.enter(span, store)
+    })
+
+    this.addSubs('replace', ([bucket]) => {
+      const store = storage.getStore()
+      const span = this.startSpan('replace', {}, store, bucket)
+      this.enter(span, store)
+    })
+
+    this.addSubs('append', ([bucket]) => {
+      const store = storage.getStore()
+      const span = this.startSpan('append', {}, store, bucket)
+      this.enter(span, store)
+    })
+
+    this.addSubs('prepend', ([bucket]) => {
+      const store = storage.getStore()
+      const span = this.startSpan('prepend', {}, store, bucket)
+      this.enter(span, store)
+    })
+  }
 }
 
-function wrapCallback (span, callback) {
-  return function (err, result) {
-    span.setTag('error', err)
-    span.finish()
+function defaultAsyncEnd () {
+  storage.getStore().span.finish()
+}
 
-    return callback.apply(this, arguments)
-  }
+function errorHandler (error) {
+  storage.getStore().span.setTag('error', error)
```

Comment:
Will a span always be available here? Are all places that call this properly bound?

Reply:
Seems like it is, because when I comment out the body of the functions outside of the plugin object, the tests fail